src/controls/LevelingSensor/QskLevelingSensorNodes.h

```diff
@@ -0,0 +1,303 @@
+
+        if (dirty)
+        {
+            update(tickmarks);
+            markDirty(QSGNode::DirtyGeometry);
+        }
+    }
+
+private:
+    void update(const QskScaleTickmarks& tickmarks)
+    {
+        if (m_geometry.vertexCount() != tickmarks.tickCount() * 2)
+        {
+            m_geometry.allocate(tickmarks.tickCount() * 2);
+        }
+
+        auto* vertexData = m_geometry.vertexDataAsPoint2D();
+
+        using T = QskScaleTickmarks::TickType;
+        for (auto type : { T::MinorTick, T::MediumTick, T::MajorTick })
+        {
+            for (const auto tick : tickmarks.ticks(type))
+            {
+                const auto i = static_cast<int>(type);
+
+                const auto p = m_scale * tick;
+                const auto d = QVector2D(-m_scale.y(), m_scale.x()).normalized();
+
+                const auto p1 = m_tickmarkSize[i] * +1 * d + p + m_offset;
+                const auto p2 = m_tickmarkSize[i] * -1 * d + p + m_offset;
+
+                vertexData[0].set(p1.x(), p1.y());
+                vertexData[1].set(p2.x(), p2.y());
+                vertexData += 2;
+            }
+        }
+
+        m_geometry.markVertexDataDirty();
+    }
+
+    QSGGeometry m_geometry;
+    QSGFlatColorMaterial m_material;
+    QVector2D m_scale = { 1.0f, 0.0f };
+    QVector2D m_offset = { 0.0f, 0.0f };
+    QVector3D m_tickmarkSize = { 1.0, 2.0, 4.0 };
+};
+
+class RadialClipNode final : public QSGClipNode
+{
+public:
+    RadialClipNode() : m_geometry(QSGGeometry::defaultAttributes_Point2D(), 0)
+    {
+        m_geometry.setVertexDataPattern(QSGGeometry::DynamicPattern);
+        m_geometry.setDrawingMode(QSGGeometry::DrawTriangleFan);
+        setGeometry(&m_geometry);
+        setIsRectangular(false);
+    }
+
+    void setGeometryProperties(const qreal radius = 1.0, const qreal cx = 0.0, const qreal cy = 0.0, const int count = 360)
+    {
+        auto dirty = false;
+        dirty |= compareExchange(m_radius, radius);
+        dirty |= compareExchange(m_cx, cx);
+        dirty |= compareExchange(m_cy, cy);
+        dirty |= compareExchange(m_count, count);
+
+        if (dirty)
+        {
+            update();
+        }
+    }
+
+private:
+    void update()
+    {
+        const auto step = 2.0 * M_PI / m_count;
+
+        if (m_geometry.vertexCount() != m_count)
+        {
+            m_geometry.allocate(m_count);
+        }
+
+        auto* vertices = m_geometry.vertexDataAsPoint2D();
+
+        for (int i = 0; i < m_count; ++i)
+        {
+            vertices[i].x = qFastCos(i * step) * m_radius + m_cx;
+            vertices[i].y = qFastSin(i * step) * m_radius + m_cy;
+        }
+
+        m_geometry.markVertexDataDirty();
+        markDirty(QSGNode::DirtyGeometry);
+    }
+
+    using QSGClipNode::setIsRectangular;
+    using QSGClipNode::setClipRect;
+
+    QSGGeometry m_geometry;
+    qreal m_radius = 1.0;
+    qreal m_cx = 0;
+    qreal m_cy = 0;
+    int m_count = 360;
+};
+
+template<typename CRTP>
```

Comment:
What does CRTP mean? Doesn't sound easy to understand...

Reply:
Curiously Recurring Template Pattern

https://en.wikipedia.org/wiki/Curiously_recurring_template_pattern